packages/gitbook/src/routes/markdownPage.ts

```diff
@@ -1,123 +1,117 @@
+    const isGroupPage = await shouldTreatAsGroupPage(context, page);
+    if (isGroupPage) {
+        return servePageGroup(context, page);
     }
 
-    // Otherwise, stream the markdown from the page and its children.
-    return new Response(
-        new ReadableStream<Uint8Array>({
-            async pull(controller) {
-                await streamMarkdownFromPage(context, page, controller);
-                controller.close();
-            },
-        }),
-        {
-            headers: {
-                'Content-Type': 'text/markdown; charset=utf-8',
-            },
-        }
+    // Handle regular document pages
+    const markdown = await throwIfDataError(
+        context.dataFetcher.getRevisionPageMarkdown({
+            spaceId: context.space.id,
+            revisionId: context.revision.id,
+            pageId: page.id,
+        })
     );
-}
-
-/**
- * Stream markdown content from a page and its children
- */
-async function streamMarkdownFromPage(
-    context: GitBookSiteContext,
-    page: RevisionPageDocument,
-    stream: ReadableStreamDefaultController<Uint8Array>
-): Promise<void> {
-    const mainPageMarkdown = await fetchMarkdown(context, page);
-    stream.enqueue(new TextEncoder().encode(mainPageMarkdown));
 
-    if (page.pages.length > 0) {
-        await streamChildPages(context, page.pages, stream);
-    }
+    return new Response(markdown, {
+        headers: {
+            'Content-Type': 'text/markdown; charset=utf-8',
+        },
+    });
 }
 
 /**
- * Stream markdown from child pages with controlled concurrency.
- * This function recursively handles nested children by streaming them as they become available.
+ * Determine if a page should be treated as a group page.
+ * A page is treated as a group if:
+ * 1. It's explicitly a group page type, OR
+ * 2. It's a document page but has empty content (acts as a container)
  */
-async function streamChildPages(
+async function shouldTreatAsGroupPage(
     context: GitBookSiteContext,
-    pages: RevisionPage[],
-    stream: ReadableStreamDefaultController<Uint8Array>
-): Promise<void> {
-    const eligiblePages = getEligiblePages(pages);
-
-    const childPagesMarkdown = pMapIterable(
-        eligiblePages,
-        async (childPage) => fetchMarkdown(context, childPage),
-        {
-            concurrency: MAX_CONCURRENCY,
-        }
-    );
-
-    for await (const childMarkdown of childPagesMarkdown) {
-        stream.enqueue(new TextEncoder().encode(`\n\n${childMarkdown}`));
+    page: RevisionPageDocument | RevisionPageGroup
+): Promise<boolean> {
+    if (page.type === RevisionPageType.Group) {
+        return true;
     }
-}
 
-/**
- * Fetch markdown from a page.
- */
-async function fetchMarkdown(
-    context: GitBookSiteContext,
-    page: RevisionPageDocument
-): Promise<string> {
-    const markdown = await throwIfDataError(
-        context.dataFetcher.getRevisionPageMarkdown({
+    const document = await throwIfDataError(
```

Comment:
Hum this is actually going to be quite expensive, as it means for each page, we'll fetch the content of the page twice, once as JSON, once as Markdown.

Can't we detect the "empty" case from the actual markdown fetched?